ext/dom/html5_serializer.c

```diff
@@ -0,0 +1,351 @@
+static zend_result dom_html5_escape_string(dom_html5_serialize_context *ctx, const char *content, bool attribute_mode)
+{
+	const char *last_output = content;
+
+	while (*content != '\0') {
+		switch (*content) {
+			/* Step 1 */
+			case '&': {
+				TRY(ctx->write_string_len(ctx->application_data, last_output, content - last_output));
+				TRY(ctx->write_string_len(ctx->application_data, "&amp;", strlen("&amp;")));
+				last_output = content + 1;
+				break;
+			}
+
+			/* Step 2 (non-breaking space) (note: uses UTF-8 internally) */
+			case '\xC2': {
+				if (content[1] == '\xA0') {
+					TRY(ctx->write_string_len(ctx->application_data, last_output, content - last_output));
+					TRY(ctx->write_string_len(ctx->application_data, "&nbsp;", strlen("&nbsp;")));
+					content++; /* Consume A0 too */
+					last_output = content + 1;
+				}
+				break;
+			}
+
+			/* Step 3 */
+			case '"': {
+				if (attribute_mode) {
+					TRY(ctx->write_string_len(ctx->application_data, last_output, content - last_output));
+					TRY(ctx->write_string_len(ctx->application_data, "&quot;", strlen("&quot;")));
+					last_output = content + 1;
+				}
+				break;
+			}
+
+			/* Step 4 */
+			case '<': {
+				if (!attribute_mode) {
+					TRY(ctx->write_string_len(ctx->application_data, last_output, content - last_output));
+					TRY(ctx->write_string_len(ctx->application_data, "&lt;", strlen("&lt;")));
+					last_output = content + 1;
+				}
+				break;
+			}
+			case '>': {
+				if (!attribute_mode) {
+					TRY(ctx->write_string_len(ctx->application_data, last_output, content - last_output));
+					TRY(ctx->write_string_len(ctx->application_data, "&gt;", strlen("&gt;")));
+					last_output = content + 1;
+				}
+				break;
+			}
+		}
+
+		content++;
+	}
+
+	return ctx->write_string_len(ctx->application_data, last_output, content - last_output);
+}
+
+static zend_result dom_html5_serialize_text_node(dom_html5_serialize_context *ctx, const xmlNode *node)
+{
+	if (node->parent->type == XML_ELEMENT_NODE && dom_is_html_ns(node->parent)) {
+		const xmlNode *parent = node->parent;
+		size_t name_length = strlen((const char *) parent->name);
+		/* Note: <noscript> is not handled because scripting is not enabled because the user agent (PHP) does not support (JS) scripting */
+		if (dom_local_name_compare_ex(parent, "style", strlen("style"), name_length)
+			|| dom_local_name_compare_ex(parent, "script", strlen("script"), name_length)
+			|| dom_local_name_compare_ex(parent, "xmp", strlen("xmp"), name_length)
+			|| dom_local_name_compare_ex(parent, "iframe", strlen("iframe"), name_length)
+			|| dom_local_name_compare_ex(parent, "noembed", strlen("noembed"), name_length)
+			|| dom_local_name_compare_ex(parent, "noframes", strlen("noframes"), name_length)
+			|| dom_local_name_compare_ex(parent, "plaintext", strlen("plaintext"), name_length)) {
+			return ctx->write_string(ctx->application_data, (const char *) node->content);
+		}
```

Comment:
I might not understand what the comment/code is doing.

But is this function ignoring the script tags because we don't support scripting and we process noscript or not?